docs/adapter_impl_notes.md

```diff
@@ -0,0 +1,184 @@
+
+For this document, we'll use the Gmail Adapter as an example. In this case, the backend consists a set of email messages. Each message contains a timestamp (when the message was received) and fields (email headers such as ``From:``, ``To:``, ``Subject:``, the message body, etc). The Gmail API supports [search expressions](https://support.google.com/mail/answer/7190?hl=en) that select messages based on date, headers, or a free-text search on the message contents.
+
+As of 2015-12-23, the adapter implementation doesn't *exactly* match this description. For example, the adapter doesn't yet support filtering expressions or live programs. However, the notes still do a good job of illustrating the concepts.
+
+The adapter's job is to interpret the options included in the juttle ``read`` command into a set of matching email messages, construct json objects representing those messages, and pass them to the Juttle Runtime by calling ``emit()``.
+
+More sophisticated adapters work together with the juttle optimizer to push aggregation operations directly into the backend. For example, to count the number of messages in a given time period you could simply fetch all the messages and have the juttle program perform the counting. But it would be more efficient to count the number of messages via Gmail APIs and simply return the count instead.
+
+## Live vs Historical
+
+Juttle's ``read`` command specifies a timerange based on the values of the ``-from`` and ``-to`` options. When the ``-to`` option is less than the current time, an adapter need only fetch the matching set of messages for the provided timerange. Such a program is called *historical*. When the ``-to`` option is greater than the current time, the adapter must also watch for new messages and pass them to the program. Such a problem is called *live*. A program can be both live and historical, with a ``-from`` in the past and a ``-to`` in the future. In that case, the adapter must fetch both old and newly-arriving messages.
+
+Although Gmail messages have a natural time values (the time the message was received), not all backends do. If your backend does not have meaningful time values, but was given a timerange in the ``read`` proc, your adapter should return an error.
+
+## Fields and Searching
+
+Every ``read`` command can contain a [filter expression](https://github.com/juttle/juttle/blob/master/docs/concepts/filtering.md) that is used as an initial filter for the messages selected by the adapter. An adapter is not obligated to support a filter expression. The ``filter`` proc allows for filtering of points within programs. However, for performance reasons it is highly recommended that adapters implement filter expressions and push filtering into the backend whenever possible.
+
+A filter expression either takes the form of a free-text search or a field match expression, possibly combined with logical operators like ``AND``, ``OR``, etc.
```

Comment:
The juttle docs say "full-text search" not "free-text", I remember some discussion back in the day that arrived at this terminology.